ml4gw/waveforms/taylorf2.py

```diff
@@ -0,0 +1,165 @@
+import torch
+from torchtyping import TensorType
+
+GAMMA = 0.577215664901532860606512090082402431
+"""Euler-Mascheroni constant. Same as lal.GAMMA"""
+
+MSUN_SI = 1.988409870698050731911960804878414216e30
+"""Solar mass in kg. Same as lal.MSUN_SI"""
+
+MTSUN_SI = 4.925490947641266978197229498498379006e-6
+"""1 solar mass in seconds. Same value as lal.MTSUN_SI"""
+
+PI = 3.141592653589793238462643383279502884
+"""Archimedes constant. Same as lal.PI"""
+
+MPC_SEC = 1.02927125e14
+"""
+1 Mpc in seconds.
+"""
+
+
+def taylorf2_phase(
+    f: TensorType,
+    mass1: TensorType,
+    mass2: TensorType,
+) -> TensorType:
+    """
+    Calculate the inspiral phase for the IMRPhenomD waveform.
+    """
+    mass1_s = mass1 * MTSUN_SI
+    mass2_s = mass2 * MTSUN_SI
+    M_s = mass1_s + mass2_s
+    eta = mass1_s * mass2_s / M_s / M_s
+
+    Mf = (f.T * M_s).T
+
+    v0 = torch.ones_like(Mf)
+    v1 = (PI * Mf) ** (1.0 / 3.0)
+    v2 = v1 * v1
+    v3 = v2 * v1
+    v4 = v3 * v1
+    v5 = v4 * v1
+    v6 = v5 * v1
+    v7 = v6 * v1
+    logv = torch.log(v1)
+    v5_logv = v5 * logv
+    v6_logv = v6 * logv
+
+    # Phase coeffeciencts from https://git.ligo.org/lscsoft/lalsuite/-/blob/master/lalsimulation/lib/LALSimInspiralPNCoefficients.c  # noqa E501
+    pfaN = 3.0 / (128.0 * eta)
+    pfa_v0 = 1.0
+    pfa_v0 *= pfaN
+    pfa_v1 = 0.0
+    pfa_v1 *= pfaN
+    pfa_v2 = 5.0 * (74.3 / 8.4 + 11.0 * eta) / 9.0
+    pfa_v2 *= pfaN
+    pfa_v3 = -16.0 * PI
+    pfa_v3 *= pfaN
+    pfa_v4 = (
+        5.0
+        * (3058.673 / 7.056 + 5429.0 / 7.0 * eta + 617.0 * eta * eta)
+        / 72.0
+    )
+    pfa_v4 *= pfaN
+    pfa_v5logv = 5.0 / 3.0 * (772.9 / 8.4 - 13.0 * eta) * PI
+    pfa_v5logv *= pfaN
+    pfa_v5 = 5.0 / 9.0 * (772.9 / 8.4 - 13.0 * eta) * PI
+    pfa_v5 *= pfaN
+    pfa_v6logv = -684.8 / 2.1
+    pfa_v6 = (
+        11583.231236531 / 4.694215680
+        - 640.0 / 3.0 * PI * PI
+        - 684.8 / 2.1 * GAMMA
+        + eta * (-15737.765635 / 3.048192 + 225.5 / 1.2 * PI * PI)
+        + eta * eta * 76.055 / 1.728
+        - eta * eta * eta * 127.825 / 1.296
+        + pfa_v6logv * torch.log(torch.tensor(4.0))
+    )
+    pfa_v6logv *= pfaN
+    pfa_v6 *= pfaN
+    pfa_v7 = PI * (
+        770.96675 / 2.54016 + 378.515 / 1.512 * eta - 740.45 / 7.56 * eta * eta
+    )
+    pfa_v7 *= pfaN
+    # construct power series
+    phasing = (v7.T * pfa_v7).T
+    phasing += (v6.T * pfa_v6 + v6_logv.T * pfa_v6logv).T
+    phasing += (v5.T * pfa_v5 + v5_logv.T * pfa_v5logv).T
+    phasing += (v4.T * pfa_v4).T
+    phasing += (v3.T * pfa_v3).T
+    phasing += (v2.T * pfa_v2).T
+    phasing += (v1.T * pfa_v1).T
+    phasing += (v0.T * pfa_v0).T
+    # Divide by 0PN v-dependence
+    phasing /= v5
```

Comment:
Just a question for my own benefit: at what power of v do things become post-Newtonian?

Reply:
The newtonian term in phasing is just 3/128 * eta * v^-5. This is called 0PN or Newtonian. So think of the expansion as [Newtonian] * [1 + (0.5 PN) * v + (1 PN) * v^2 + ...].